### ISSUE TITLE
Unexpected unused variable warning in subclass method since v2.8.2

### ISSUE THREAD

**Issue by danielbachhuber** [sirbrillig/phpcs-variable-analysis#182]
I have a base class like this:

```php
abstract class Shortcode {

	abstract public static function callback( $attrs, $content = '' );
}
```

And a subclass like this:

```php
class Thumbnail_Gallery extends Shortcode {

	public static function callback( $attrs, $content = '' ) {
		return '';
	}
}
```

Since v2.8.2, `$attrs` and `$content` are (unexpectedly?) flagged as unused variables. However, removing them causes a fatal error.

**Comment by sirbrillig** [comment]
That's an interesting case. Those variables have to be there, since it's matching the overridden function, however, they are definitely unused. 2.8.2 fixed a pretty major bug which was missing a bunch of unused variable warnings, which is why you're seeing them now.

It would be nice if this sniff could ignore those variables since they're required to be unused in this instance, but doing so would require a certain amount of static analysis which is beyond the scope of most linting; we'd have to know that the function was overriding another function with the same signature.

I'm open to ideas but in this case what I'd do is just put something like `// phpcs:ignore VariableAnalysis -- function signature requires these variables` on the function definition line.

**Comment by danielbachhuber** [comment]
> 2.8.2 fixed a pretty major bug which was missing a bunch of unused variable warnings, which is why you're seeing them now.

Makes sense.

> It would be nice if this sniff could ignore those variables since they're required to be unused in this instance, but doing so would require a certain amount of static analysis which is beyond the scope of most linting; we'd have to know that the function was overriding another function with the same signature.
> 
> I'm open to ideas but in this case what I'd do is just put something like `// phpcs:ignore VariableAnalysis -- function signature requires these variables` on the function definition line.

Fair enough. Honestly, the `abstract public static function callback( $attrs, $content = '' );` isn't really necessary, so I think I'll fix it by removing the `abstract`.

**Comment by danielbachhuber** [comment]
> Fair enough. Honestly, the `abstract public static function callback( $attrs, $content = '' );` isn't really necessary, so I think I'll fix it by removing the `abstract`.

As it turns out, this introduces another problem:

```
Warning: Declaration of Thumbnail_Gallery::callback($attrs, $content = '') should be compatible with Shortcode::callback()
```

So maybe this needs to be considered in a little more depth?

**Comment by sirbrillig** [comment]
Sometimes I think we just have to explicitly say in the code "I know these variables are unused, but I'm ok with that". Here's an example: https://github.com/sirbrillig/phpcs-changed/blob/2313396feaf5ef16b3229045f27b752a814bdbb5/tests/GitWorkflowTest.php#L22

**Comment by sirbrillig** [comment]
That said, if you can think of a way to prevent this, I'm very open!

**Comment by danielbachhuber** [comment]
> Sometimes I think we just have to explicitly say in the code "I know these variables are unused, but I'm ok with that". Here's an example:

Hm, yeah but the annoying thing is that you have to do it on every subclass method.

> That said, if you can think of a way to prevent this, I'm very open!

Nothing great at the moment! I don't know the codebase for this library enough to offer suggestions, unfortunately.

**Comment by jrfnl** [comment]
The only thing I can think of for this - possibly behind a property toggle -, is to check when an unused function parameter is found, whether the class extends another and/or implements and interface and if so, not report it or only report it when the property toggle has been set to do so.

**Comment by sirbrillig** [comment]
@danielbachhuber just a thought: ...do the variable names have to match those in the parent class? I don't think so, so you could name them something like `$_ignore1` and then use `ignoreUnusedRegexp` to ignore variables named `/^_ignore/`.

**Comment by sirbrillig** [comment]
Another option is to set `allowUnusedFunctionParameters` to true.

**Comment by danielbachhuber** [comment]
> do the variable names have to match those in the parent class? I don't think so, so you could name them something like `$_ignore1` and then use `ignoreUnusedRegexp` to ignore variables named `/^_ignore/`.

@sirbrillig This approach does work, yes. However, I think it's a little bit messy :/ 

> Another option might be to set `allowUnusedFunctionParameters` to true.

Could do, but I do want to know about genuinely unused function parameters that aren't subject to the inheritance requirement.

---

For now, removing the abstract method is the easiest/best solution because it wasn't even necessary in the first place.

Thanks for your help exploring this!
